### PR TITLE
Feat/classification

### PR DESCRIPTION
**Pull request by adityajain07**

### CODE REVIEW

File: src/classification/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    is_training: bool = False,
+) -> torch.utils.data.DataLoader:
+    """Main dataset builder and loader function"""
+
+    # Load the webdataset
+    if is_training:
+        dataset = wds.WebDataset(sharedurl, shardshuffle=True)
+        dataset = dataset.shuffle(10000)
+    else:
+        dataset = wds.WebDataset(sharedurl, shardshuffle=False)
```
Comment (f-PLT) [inline src/classification/dataloader.py:99]: This also can be refactored:

```python
dataset = wds.WebDataset(sharedurl, shardshuffle=is_training)
if is_training:
    dataset = dataset.shuffle(10000)
```

**Review by f-PLT — COMMENTED** [review]
Did a first pass on this PR.

Overall, very good and it's nice to see the progress between PRs 😄  

Comments are mostly suggestions to help you see other patterns when writing if/elif/else and how they can help with maintainability and reducing duplication.

File: src/classification/constants.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+VIT_B16_384 = "vit_base_patch16_384"
+
+ADAMW = "adamw"
+SGD = "sgd"
+
+CROSS_ENTROPY_LOSS = "cross_entropy"
+
+COSINE_LR_SCHEDULER = "cosine"
+
+AVAILABLE_MODELS = frozenset(
```
[the remark this answers is not in the document]
Comment (adityajain07) [reply src/classification/constants.py:17]: Nice catch! Fixed in [3b4bc4c](https://github.com/RolnickLab/ami-ml/pull/49/commits/3b4bc4c44a5682b47fe96d21186e58292da8d055).

File: src/classification/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+def _normalization(preprocess_mode: str) -> tuple[list[float], list[float]]:
+    """Get the mean and std for normalization"""
+
+    if preprocess_mode == "torch":
+        mean, std = [0.485, 0.456, 0.406], [0.229, 0.224, 0.225]
+    elif preprocess_mode == "tf":
+        mean, std = [0.5, 0.5, 0.5], [0.5, 0.5, 0.5]
+    else:
+        mean, std = [0.0, 0.0, 0.0], [1.0, 1.0, 1.0]
```
[the remark this answers is not in the document]
Comment (adityajain07) [reply src/classification/dataloader.py]: Fixed in[659e483](https://github.com/RolnickLab/ami-ml/pull/49/commits/659e48384301cd1ceea4b3f7313c997faa4c1ec4)!

File: src/classification/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+    random_num = np.random.uniform()
+    if random_num <= 0.25:
+        transform = transforms.Resize((int(0.5 * full_size), int(0.5 * full_size)))
+        image = transform(image)
+    elif random_num <= 0.5:
+        transform = transforms.Resize((int(0.25 * full_size), int(0.25 * full_size)))
+        image = transform(image)
+
+    return image
```
[the remark this answers is not in the document]
Comment (adityajain07) [reply src/classification/dataloader.py]: This is so cool! I would have never thought about this.

Fixed in [659e483](https://github.com/RolnickLab/ami-ml/pull/49/commits/659e48384301cd1ceea4b3f7313c997faa4c1ec4)!

**Comment by f-PLT** [comment]
Don't forget to take the PR out of Draft mode and re-request a review once it's done


image

File: src/classification/utils.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+def get_loss_function(
+    loss_function_name: str, label_smoothing: float = 0.0
+) -> torch.nn.Module:
+    """Loss function definitions"""
+
+    if loss_function_name == CROSS_ENTROPY_LOSS:
+        return torch.nn.CrossEntropyLoss(label_smoothing=label_smoothing)
+    else:
+        raise RuntimeError(f"{loss_function_name} loss is not implemented.")
```
Comment (f-PLT) [inline src/classification/utils.py:94]: Ideally, these elses should also be treated like the other examples.

For example:
```python
if lr_scheduler_type != COSINE_LR_SCHEDULER:
    raise RuntimeError()

scheduler = CosineLRScheduler(
                optimizer,
                t_initial=(total_steps - warmup_steps),
                warmup_t=warmup_steps,
                warmup_prefix=True,
                cycle_limit=1,
                t_in_epochs=False,
    )
return scheduler
```

That way, your code is already ready for more cases (other scheduler types or loss functions), with no extra cost or complexity in the code

**Review by f-PLT — APPROVED** [review]
Looks good to me!

Added a small comment that is not critical.

Great job!

File: src/classification/utils.py (hunk 5 above)
Comment (adityajain07) [reply src/classification/utils.py:94]: Thanks, Francis, for this comment! I'll update it in another PR, otherwise, I won't be able to merge this one.
